scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -147,18 +157,23 @@ private[scio] class KryoAtomicCoder[T] extends AtomicCoder[T] {
   private def kryoEncodedElementByteSize(obj: Any): Long = {
     val s = new CountingOutputStream(ByteStreams.nullOutputStream())
     val output = new Output(s)
-    kryo.get().writeClassAndObject(output, obj)
+    kryoState.get().kryo.writeClassAndObject(output, obj)
     output.flush()
     s.getCount + VarInt.getLength(s.getCount)
   }
 
 }
 
+/** Used for sharing Kryo instance and buffers */
+private[scio] final case class KryoState(kryo: Kryo, input: InputChunked, output: OutputChunked)
+
 private[scio] object KryoAtomicCoder {
   def apply[T]: Coder[T] = new KryoAtomicCoder[T]
 
-  private val kryo: ThreadLocal[Kryo] = new ThreadLocal[Kryo] {
-    override def initialValue(): Kryo = {
+  val bufferSize = 64 * 1024
+
+  private val kryoState: ThreadLocal[KryoState] = new ThreadLocal[KryoState] {
```

Comment:
This will cause IOOB exceptions dues to kryo state being shared across multiple instances (within the same thread). Beam has an internal heuristic to monitor size of collections and that seems to be using causing IOOB if kryo is shared to serialize and deserialize. This was the reason for #829.